### PR TITLE
Fix RepartitionedHybridTopologyFactory tests to reduce stochastic failures

### PR DESCRIPTION
**Pull request by zhang-ivy**
I’ve added a fix to the RepartitionedHybridTopologyFactory tests to reduce stochastic failures — we now check the potential energy of the proposed positions before generating the `RepartitonedHybridTopologyFactory`. If the potential energy is greater than 1e8, we try to generate another proposal with energy < 1e8. We do this 5x, and if on the 5th time we still fail to generate a proposal with energy < 1e8, we raise an exception.

Addresses: https://github.com/choderalab/perses/issues/815

### CODE REVIEW

**Comment by mikemhenry** [comment]
Looks like we are still getting a failure:
test_flattenedHybridTopologyFactory_energies
https://github.com/choderalab/perses/pull/828/checks?check_run_id=3196637909#step:9:353

**Comment by mikemhenry** [comment]
I'm also worried that this strategy is just re-rolling the randomness of the way we are generating the atom map.

**Comment by zhang-ivy** [comment]
@mikemhenry: The failure you linked above is in the nightly -- which I thought wasn't required for merging? Also upon further inspection of the failure, it looks like the code that is being run in that version doesn't contain my changes in this PR, check the lines below [here](https://github.com/choderalab/perses/pull/828/checks?check_run_id=3196637909#step:9:296) -- they don't contain the for loop I introduced

**Comment by zhang-ivy** [comment]
> I'm also worried that this strategy is just re-rolling the randomness of the way we are generating the atom map.

I don't think its randomness in the atom map that's causing the stochastic failures, I think its randomness in the geometry engine's proposal order. The geometry engine by nature was designed to use randomness to generate new proposals, so I don't think that part would change. If anyone has any suggestions for alternative ways to solve this, feel free to share here! I just implemented a version of what @jchodera  suggested [here](https://github.com/choderalab/perses/pull/809#issuecomment-886260884)

**Review by jchodera — APPROVED** [review]
This is certainly a significant improvement!
We can continue to improve this to be more deterministic and thorough in future iterations.

**Comment by zhang-ivy** [comment]
@mikemhenry : I think you are putting out a release today -- feel free to merge this and include it in the release if you see fit!

**Comment by mikemhenry** [comment]
That sounds good, I will get it into the release today!

RE: that failure not containing the PR changes: I will investigate that since we should be testing the code in the PR, thanks for the great work @zhang-ivy

File: perses/tests/test_relative.py [hunk 1]
```diff
@@ -711,10 +712,25 @@ def RepartitionedHybridTopologyFactory_energy_mutant(topology, chain, system, po
...
+        context.setPeriodicBoxVectors(*topology_proposal.new_system.getDefaultPeriodicBoxVectors())
+        context.setPositions(new_positions)
+        context.setVelocitiesToTemperature(temperature)
+        potential_energy = context.getState(getEnergy=True).getPotentialEnergy()
+
+        if potential_energy < 1e8 * unit.kilojoule_per_mole:
+            break
+        else:
+            if i == 4:
+                raise Exception("Was not able to generate a decent proposal after 5 iterations")
```
Comment (ijpulidos) [inline perses/tests/test_relative.py:733]: This is nice! I believe a more readable way would be using a `for-else` statement here. I will be making the changes to accomplish that.

**Review by ijpulidos — COMMENTED** [review]
Nice change, this will definitely help with the issue.

File: perses/tests/test_relative.py (hunk 1 above)
Comment (ijpulidos) [reply perses/tests/test_relative.py]: Fixed in https://github.com/choderalab/perses/pull/828/commits/867b40e48d249213a83349d28b4123fe7c8637e1

Comment (zhang-ivy) [reply perses/tests/test_relative.py]: Cool, thanks! I didn't know about `for-else` statements!

**Comment by zhang-ivy** [comment]
@mikemhenry @ijpulidos : It looks like one of the required tests is now failing. When I inspect the error, it again looks like this PR is not being tested properly. There should be a for loop starting around [here](https://github.com/choderalab/perses/pull/828/checks?check_run_id=3203785416#step:9:312).

**Comment by mikemhenry** [comment]
It is because we need to copy this fix to another test as well

**Comment by mikemhenry** [comment]
In the future we can refactor things to use fixtures so we can re-use bits of code for testing instead of copying and pasting the same bits in each test, this will make maintenance like this much easier!

**Comment by zhang-ivy** [comment]
>It is because we need to copy this fix to another test as well

Ah good catch! I had forgotten about the other test 

>In the future we can refactor things to use fixtures so we can re-use bits of code for testing instead of copying and pasting the same bits in each test, this will make maintenance like this much easier!

This sounds great!

**Comment by mikemhenry** [comment]
I copied the fix, so once it passes we should be good!

**Review by mikemhenry — APPROVED** [review]

**Pull request merged**